src/types/deleted.rs

```diff
@@ -0,0 +1,259 @@
+use super::{Seq, Uid};
+use std::ops::RangeInclusive;
+
+/// An enum representing message sequence numbers or UID sequence sets returned
+/// in response to a `EXPUNGE` command.
+///
+/// The `EXPUNGE` command may return several `EXPUNGE` responses referencing
+/// message sequence numbers, or it may return a `VANISHED` response referencing
+/// multiple UID values in a sequence set if the client has enabled
+/// [QRESYNC](https://tools.ietf.org/html/rfc7162#section-3.2.7).
+///
+/// `Deleted` implements some iterators to make it easy to use. If the caller
+/// knows that they should be receiving an `EXPUNGE` or `VANISHED` response,
+/// then they can use [`seqs()`](#method.seqs) to get an iterator over `EXPUNGE`
+/// message sequence numbers, or [`uids()`](#method.uids) to get an iterator over
+/// the `VANISHED` UIDs. As a convenience `Deleted` also implents `IntoIterator`
+/// which just returns an iterator over whatever is contained within.
+///
+/// # Examples
+/// ```no_run
+/// # let domain = "imap.example.com";
+/// # let tls = native_tls::TlsConnector::builder().build().unwrap();
+/// # let client = imap::connect((domain, 993), domain, &tls).unwrap();
+/// # let mut session = client.login("name", "pw").unwrap();
+/// // Iterate over whatever is returned
+/// if let Ok(deleted) = session.expunge() {
+///     for id in &deleted {
+///         // Do something with id
+///     }
+/// }
+///
+/// // Expect a VANISHED response with UIDs
+/// if let Ok(deleted) = session.expunge() {
+///     if let Some(uid_iter) = deleted.uids() {
+///         for uid in uid_iter {
+///             // Do something with uid
+///         }
+///     }
+/// }
+/// ```
+#[derive(Debug, Clone)]
+pub enum Deleted {
+    /// Message sequence numbers given in an `EXPUNGE` response.
+    Expunged(Vec<Seq>),
+    /// Message UIDs given in a `VANISHED` response.
+    Vanished(Vec<RangeInclusive<Uid>>),
+}
+
+impl<'a> Deleted {
+    /// Construct a new `Deleted` value from a vector of message sequence
+    /// numbers returned in one or more `EXPUNGE` responses.
+    pub fn from_expunged(v: Vec<u32>) -> Self {
+        Deleted::Expunged(v)
+    }
+
+    /// Construct a new `Deleted` value from a sequence-set of UIDs
+    /// returned in a `VANISHED` response
+    pub fn from_vanished(v: Vec<RangeInclusive<u32>>) -> Self {
+        Deleted::Vanished(v)
+    }
+
+    /// Return an iterator over message sequence numbers from an `EXPUNGE`
+    /// response. If the client is expecting sequence numbers this function
+    /// can be used to ensure only sequence numbers returned in an `EXPUNGE`
+    /// response are processed.
+    pub fn seqs(&'a self) -> Option<DeletedIterator<'a>> {
```

Comment:
I think you could probably make this and `uids` much nicer by returning `impl ExactSizeIterator<Item = u32>`. That way you can get rid of all the various `struct` types and `impl Iterator for` blocks below, and just use combinators instead (`flat_map` will probably come in handy for `uids` for example).

I also don't think the `Option` adds much here. `ExactSizeIterator` is good enough since it also provides an `is_empty` method.

Reply:
Oh, this is much better - thanks! I had not thought to do it this way.

I pushed a new version with `seqs()` and `uids()` returning `impl Iterator<Item = u32>`. `ExactSizeIterator` isn't implemented for `FlatMap` (which makes sense, I suppose), so I used `Iterator` instead. I could use `ExactSizeIterator` for `seqs()`, but wanted to keep them consistent.

As for the `Option` - I included it to let the caller distinguish the 'no values' from the 'didn't get that kind of response' cases, but wasn't sure if this was a distinction that added enough value to justify the `Option`, so am glad you have an opinion! I removed it in this version, though `Iterator` doesn't provide `is_empty()` like `ExactSizeIterator` does, but I am not entirely convinced this is needed? `Deleted` has `is_empty()` anyway. If you think this is important I can figure it out.

Thanks for the comments - I found all the structs a bit cumbersome, but didn't think to just return an `impl`. Being able to delete all that makes it much simpler / nicer.

Reply:
Also, TIL `flat_map()` is a thing. That is much nicer than rolling it by hand!